### PR TITLE
EVM-C: Change get_code() to copy_code()

### PR DESCRIPTION
**Pull request by chfast**
Deprecates #151.

### CODE REVIEW

File: libevmjit/JIT.cpp [hunk 1]
```diff
@@ -230,6 +232,19 @@ int64_t call_v2(
 }
 
 
+/// A wrapper for new EVM-C copycode callback function.
+size_t extcode(uint8_t** o_pCode, evm_context* _ctx, evm_address const* _address) noexcept
+{
+    auto& jit = JITImpl::instance();
+    jit.codeBuffer.resize(0x6000);  // The max contract code size.
```
Comment (axic) [inline libevmjit/JIT.cpp]: This is hard fork dependant. It was only introduced in spurious dragon?

Comment (chfast) [reply libevmjit/JIT.cpp]: Yes. You can make a test for pre Spurious Dragon that will fail here.

Comment (axic) [reply libevmjit/JIT.cpp]: How about using `.resize(host->code_size())` ?

File: libevmjit/JIT.cpp [hunk 2]
```diff
@@ -250,7 +266,7 @@ class SymbolResolver : public llvm::SectionMemoryManager
 			.Case("evm.sstore", reinterpret_cast<uint64_t>(jit.host->set_storage))
 			.Case("evm.balance", reinterpret_cast<uint64_t>(jit.host->get_balance))
 			.Case("evm.codesize", reinterpret_cast<uint64_t>(jit.host->get_code_size))
-			.Case("evm.code", reinterpret_cast<uint64_t>(jit.host->get_code))
+			.Case("evm.code", reinterpret_cast<uint64_t>(extcode))
```
Comment (axic) [inline libevmjit/JIT.cpp]: Should this actually be named `extcodecopy` or rather `get_code`?

File: include/evm.h [hunk 3]
```diff
@@ -351,20 +351,27 @@ typedef void (*evm_get_balance_fn)(struct evm_uint256be* result,
...
-/// Get code callback function.
+/// Copy code callback function.
 ///
-/// This callback function is used by an EVM to get the code of a contract of
-/// given address.
+/// This callback function is used by an EVM to request a copy of the code
+/// of the given account to the memory buffer provided by the EVM.
+/// The Client MUST copy the requested code, starting with the given offset,
+/// to the provided memory buffer up to the size of the buffer or the size of
+/// the code, whatever is smaller.
```
Comment (axic) [inline include/evm.h]: whatever -> whichever

File: examples/capi.c [hunk 4]
```diff
@@ -74,11 +74,10 @@ static size_t get_code_size(struct evm_context* context, const struct evm_addres
     return 0;
 }
 
-static size_t get_code(const uint8_t** code,
-                       struct evm_context* context,
-                       const struct evm_address* address)
+static size_t copy_code(struct evm_context* context, const struct evm_address* address,
+    size_t code_offset, uint8_t* output_data, size_t output_size)
```
Comment (axic) [inline examples/capi.c]: Please use the same parameter names as in the header (`buffer`, `buffer_size`).

File: libevmjit/JIT.cpp [hunk 5]
```diff
@@ -243,22 +259,23 @@ class SymbolResolver : public llvm::SectionMemoryManager
...
-			return {addr, llvm::JITSymbolFlags::Exported};
+        auto addr =
+            llvm::StringSwitch<uint64_t>(unprefixedName)
+                .Case("env_sha3", reinterpret_cast<uint64_t>(&keccak))
+                .Case("evm.exists", reinterpret_cast<uint64_t>(jit.host->account_exists))
+                .Case("evm.sload", reinterpret_cast<uint64_t>(jit.host->get_storage))
+                .Case("evm.sstore", reinterpret_cast<uint64_t>(jit.host->set_storage))
+                .Case("evm.balance", reinterpret_cast<uint64_t>(jit.host->get_balance))
+                .Case("evm.codesize", reinterpret_cast<uint64_t>(jit.host->get_code_size))
+                .Case("evm.code", reinterpret_cast<uint64_t>(getCode))
```
Comment (axic) [inline libevmjit/JIT.cpp]: Don't want to be too picky, but other functions use underscores (`call_v2`)? :)

File: examples/capi.c [hunk 6]
```diff
@@ -74,11 +74,10 @@ static size_t get_code_size(struct evm_context* context, const struct evm_addres
     return 0;
 }
 
-static size_t get_code(const uint8_t** code,
-                       struct evm_context* context,
-                       const struct evm_address* address)
+static size_t copy_code(struct evm_context* context, const struct evm_address* address,
+    size_t code_offset, uint8_t* buffer_data, size_t buffer_size)
```
Comment (axic) [inline examples/capi.c]: The header uses `buffer`, this one uses `buffer_data`. Which one do you want to go with?

File: libevmjit/JIT.cpp (hunk 5 above)
Comment (chfast) [reply libevmjit/JIT.cpp]: Changed to camelCase.

File: examples/capi.c (hunk 6 above)
Comment (chfast) [reply examples/capi.c]: I prefer using `_data` suffix to easy spot `_data` and `_size` pairs.

**Pull request merged**